File: python/src/n0543_diameter_of_binary_tree.py
from typing import Optional

from utils.tree import TreeNode


def diameter_of_binary_tree(root: Optional[TreeNode]) -> int:
    max_diameter: int = 0

    def depth(root: Optional[TreeNode]) -> int:
        nonlocal max_diameter
        if not root:
            return 0
        left = depth(root.left)
        right = depth(root.right)

        max_diameter = max(max_diameter, left + right)

        return 1 + max(left, right)

    depth(root)

    return max_diameter

File: typescript/src/n0543_diameter_of_binary_tree.ts
import type { TreeNode } from "./utils/tree.ts";

export function diameterOfBinaryTree(root: TreeNode | null): number {
  let max_diameter = 0;

  function depth(root: TreeNode | null): number {
    if (!root) {
      return 0;
    }

    const left = depth(root.left);
    const right = depth(root.right);

    max_diameter = Math.max(max_diameter, left + right);

    return 1 + Math.max(left, right);
  }
  depth(root);

  return max_diameter;
}
